R/csv-to-duckdb.R

```diff
@@ -0,0 +1,303 @@
+
+
+#' Create a query string for loading csvs to a database.
+#'
+#' @param con database connection
+#' @param path Character vector of one or more file paths, or a single string with a globbing pattern.
+#'    E.g., `c("R/dir/a.csv", "R:/dir/b.csv")` OR `"R:/dir/*.csv"``.
+#' @param tbl_name The name of the table to write
+#' @param schema named character vector mapping column names to duckdb data types.
+#' @param col_rename a named vector of columns to rename, with names as the original
+#'    names, and values as the new names.
+#'    E.g. `c(PHNNUM = "studyid", OLDNAME2 = "newname2")`
+#' @param date_format format of date columns. Default `"%Y%m"`
+#' @param timestamp_format format of timestamp columns. Default `"%Y%m%d%H%M"`
+#' @param dt_override named list of date/datetime formats for columns
+#'    that don't follow those defined in `date_format` and `timestamp_format`.
+#'    Use this when a column has more than one date or datetime format.
+#'    eg. `list(WEIRDATE = "%m/%d/%YT%H:%M:%S", OTHERCOL = c("%m/%d/%Y %I:%M:%S %p","%m/%d/%Y"))`
+#' @param nullstr values that denote missing data. Character vector. Default `NULL` uses
+#'    duckdb defaults. E.g., `nullstr = c("NA", "")`
+#' @param opts Other options passed on to the internal `duckdb` `read_csv()`. A
+#'    list with named elements. E.g.,
+#'    `list(thousands = ",", comment = "#", force_not_null = c("a", "b"), null_padding = TRUE, buffer_size = 5)`
+#'
+#' @return a character string of the query, to be called with [DBI::dbExecute()]
+#' @export
+csv_to_duckdb_query <- function(
+    con,
+    path,
+    tbl_name,
+    schema,
+    col_rename = NULL,
+    date_format = "%Y%m",
+    timestamp_format = "%Y%m%d%H%M",
+    dt_override = NULL,
+    nullstr = NULL,
+    opts = list()
+) {
+
+  select_statement <- make_select_statement(schema, col_rename, dt_override)
+
+  sql_schema <- make_sql_schema_string(con, schema, dt_override)
+
+  path <- vec_to_string_or_array(path)
+
+  opts <- parse_opts(con, nullstr, opts)
+
+  glue::glue(
+    "
+  CREATE OR REPLACE TABLE {tbl_name} AS
+  SELECT
+    {select_statement}
+  FROM read_csv(
+    {path},
+    types = {sql_schema},
+    dateformat = '{date_format}',
+    timestampformat = '{timestamp_format}'{opts}
+  )
+  "
+  )
+}
+
+make_select_statement <- function(schema, col_rename = NULL, dt_override = NULL) {
```

Comment:
When the final SQL CREATE or REPLACE TABLE statement is executed via `DBI::dbExecute()`, column names with spaces (like First Name AS first_name) will cause SQL syntax errors. The `dipr` functions are behaving as expected, but it could be a "nice to have" to warn against this. Here's an example: 

`> print(strSQL)`
`CREATE OR REPLACE TABLE starwars AS
SELECT
  Name Abbr AS name_abbr,
  Name AS name
FROM read_csv(
  './inst/extdata/starwars-dict.csv',
  types = {
    'Name Abbr': 'VARCHAR',
    'Name': 'VARCHAR'
  })`
`> DBI::dbExecute(con,strSQL)`

This will cause `dbSendQuery()` to error.